File: Algorithms/Easy/121. Best Time to Buy and Sell Stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n=prices.size();
        if(n==0) return 0;
        vector<int> pre(n),suff(n);
        pre[0]=prices[0];
        for(int i=1;i<n;i++) {
            if(prices[i]<pre[i-1]) pre[i]=prices[i];
            else pre[i]=pre[i-1];
        }
        suff[n-1]=prices[n-1];
        for(int i=n-2;i>=0;i--) {
            if(prices[i]>suff[i+1]) suff[i]=prices[i];
            else suff[i]=suff[i+1];
        }
        
        int ans=0;
        
        for(int i=0;i<n;i++){
            if(suff[i]>pre[i])
                ans=max(ans,suff[i]-pre[i]);
        }
        return ans;
    }
};

File: Java-practice/121. Best Time to Buy and Sell Stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int ans = 0;
        int value = prices[0];
        for(int i = 1;i < prices.length;i++) {
            ans = Math.max(ans,prices[i] - value);
            value = Math.min(value,prices[i]);
        }
        return ans;
    }
}




//Recursion with Memoization
class Solution { 

    int profit(int dp[][],int prices[],int index,int buy) {
        if(index == prices.length || buy == 2) {
            return 0;
        }

        if(dp[index][buy] != -1) return dp[index][buy];

        if(buy == 1) {
            return dp[index][buy] = Math.max(-prices[index] + profit(dp,prices,index+1,0),profit(dp,prices,index+1,1));
        }
        else {
            return dp[index][buy] = Math.max(prices[index] + profit(dp,prices,index+1,2),profit(dp,prices,index+1,0));
        }
    }


    public int maxProfit(int[] prices) {
        int dp[][] = new int[prices.length][3];

        for(int row[] : dp) {
            Arrays.fill(row,-1);
        }

        return profit(dp,prices,0,1);
    }
}
